### PR TITLE
DR- 994 Exclude snapshots that are exclusively locked from lookups

### PR DESCRIPTION
**Pull request by RoriCremer**

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/BigQueryPartitionConfigV1.java [hunk 1]
```diff
@@ -14,12 +14,12 @@
...
-    @JsonProperty private final String columnName;
-    @JsonProperty private final Long intMin;
-    @JsonProperty private final Long intMax;
-    @JsonProperty private final Long intInterval;
+    @JsonProperty private long version;
+    @JsonProperty private Mode mode;
+    @JsonProperty private String columnName;
+    @JsonProperty private Long intMin;
+    @JsonProperty private Long intMax;
+    @JsonProperty private Long intInterval;
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/dataset/BigQueryPartitionConfigV1.java]: I think this is unrelated to this PR. Just curious, what was the reason for this change?

File: src/main/java/bio/terra/service/filedata/FileService.java [hunk 2]
```diff
@@ -168,12 +178,12 @@ public FileModel lookupSnapshotPath(String snapshotId, String path, int depth) {
     }
 
     FSItem lookupSnapshotFSItem(String snapshotId, String fileId, int depth) {
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/filedata/FileService.java:181]: Maybe add a comment for this method warning about it not returning snapshots that are exclusively locked?

File: src/main/java/bio/terra/service/filedata/FileService.java [hunk 3]
```diff
@@ -168,12 +178,12 @@ public FileModel lookupSnapshotPath(String snapshotId, String path, int depth) {
...
-        Snapshot snapshot = snapshotService.retrieve(UUID.fromString(snapshotId));
+        Snapshot snapshot = snapshotService.retrieveAvailable(UUID.fromString(snapshotId));
         return fileDao.retrieveById(snapshot, fileId, depth, true);
     }
 
     FSItem lookupSnapshotFSItemByPath(String snapshotId, String path, int depth) {
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/filedata/FileService.java:187]: Same here -- comment with a warning I think would be helpful.

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java [hunk 4]
```diff
@@ -99,6 +108,8 @@
     @Autowired private ConnectedOperations connectedOperations;
     @Autowired private ConnectedTestConfiguration testConfig;
     @Autowired private ConfigurationService configService;
+    @Autowired private DrsIdService drsIdService;
+    //@Autowired private EncodeFileTest encodeFileTest;
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java]: Nit: Remove if unused?

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java [hunk 5]
```diff
@@ -375,6 +386,193 @@ public void testOverlappingDeletes() throws Exception {
...
+        SnapshotSummaryModel snapshotSummary = connectedOperations.createSnapshot(
+            datasetSummary, "simple-with-filerefs-snapshot.json", "");
+
+        // check that the snapshot metadata row is unlocked
+        String exclusiveLock = snapshotDao.getExclusiveLock(UUID.fromString(snapshotSummary.getId()));
+        assertNull("snapshot row is unlocked", exclusiveLock);
+
+        // retrieve the snapshot and check that it finds it
+        SnapshotModel snapshotModel = connectedOperations.getSnapshot(snapshotSummary.getId());
+        assertEquals("Lookup unlocked snapshot succeeds", snapshotSummary.getName(), snapshotModel.getName());
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java]: I think you don't need this check, because the retrieve snapshot endpoint was tested in the previous test.

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java [hunk 6]
```diff
@@ -375,6 +386,193 @@ public void testOverlappingDeletes() throws Exception {
...
+        // lookup the snapshot file by DRS path and check that it's found
+        FileModel fsObjByPath =
+            connectedOperations.lookupSnapshotFileByPathSuccess(snapshotSummary.getId(), filePath, 0);
+        assertEquals("Retrieve snapshot file by path matches desc", fsObjByPath.getDescription(), fileLoadModel.getDescription());
+        assertThat("Retrieve snapshot file objects match", fsObjById, CoreMatchers.equalTo(fsObjByPath));
+
+        // now the snapshot exists....let's get it locked!
+
+        // enable wait in DeleteSnapshotPrimaryDataStep
+        configService.setFault(ConfigEnum.SNAPSHOT_DELETE_LOCK_CONFLICT_STOP_FAULT.name(), true);
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java]: You should move the asserts outside of the hang block here to reduce the chance of the test failing before the hang is disabled. This refactoring happened after I pointed you at the related dataset connected tests, so apologies for the late notice. You can see the changes I'm talking about in the PRs for DR-1048 and DR-994.

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java [hunk 7]
```diff
@@ -561,4 +759,30 @@ private long queryForCount(
         return countValue.getLongValue();
     }
 
+    private String getFileRefIdFromSnapshot(SnapshotSummaryModel snapshotSummary) throws InterruptedException {
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java:809]: I like pulling some of this BigQuery stuff out into a helper method, but I think it's very specific to your particular test. I wonder if you could factor out the snapshot, table and column names to make it more general? And maybe also the where/limit clause? If that's too much work, then I think you could also just put this inline into the test.

**Review by marikomedlock — COMMENTED** [review]
Added comments below. I think it would be good to rebase this PR off of DR-994, so that people aren't reviewing the same changes in both PRs. I *think* I separated out the changes you're making from DR-994, but also let me know if I've missed some.

File: src/main/java/bio/terra/service/dataset/BigQueryPartitionConfigV1.java (hunk 1 above)
Comment (ddietterich) [reply src/main/java/bio/terra/service/dataset/BigQueryPartitionConfigV1.java]: The members can only be `final` if they are filled in by the constructor. Since we have to use a default constructor for Jackson to be able to deserialize this class, we couldn't make them final anymore.

**Review by ddietterich — APPROVED** [review]
LGTM
again, maybe possible to DRY out the tests

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java (hunk 6 above)
Comment (RoriCremer) [reply src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java]: I'm wondering if this needs to be added back into:
testExcludeLockedFromDatasetLookups and testExcludeLockedFromSnapshotLookups and finally testExcludeLockedFromFileLookups

**Review by fboulnois — APPROVED** [review]
lgtm

File: src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java [hunk 8]
```diff
@@ -672,4 +798,30 @@ private long queryForCount(
...
+
+        String sql = builder.toString();
+        try {
+            QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(sql).build();
+            TableResult result = bigQueryProject.getBigQuery().query(queryConfig);
+            FieldValueList row = result.iterateAll().iterator().next();
+            FieldValue idValue = row.get(0);
+            String drsUri = idValue.getStringValue();
+            return drsUri;
+        } catch (InterruptedException ie) {
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/snapshot/SnapshotConnectedTest.java]: Do you need both this try/catch for InterruptedException and adding it to the method signature?

**Review by marikomedlock — APPROVED** [review]
LGTM! One minor comment on helper method in test.

**Comment by RoriCremer** [comment]
note---a refactor that is a planned will shorten the length of the method that is marked by Codacy as too long
